File: src/main/java/com/morak/performancetracker/description/LoggingDescriptor.java
package com.morak.performancetracker.description;

import com.morak.performancetracker.context.*;
import com.morak.performancetracker.utils.ListUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

@Component
@ConditionalOnProperty(value = "com.morak.performance-tracker.format", havingValue = "log", matchIfMissing = true)
public class LoggingDescriptor implements Descriptor {

    private final Logger log = LoggerFactory.getLogger("PERFORMANCE");

    @Override
    public void describe(Result root) {
        describeInternal(root, 0);
    }

    private void describeInternal(Result entry, int indent) {
        describeOnDepth(entry.getResult(), indent);
        if (ListUtils.isNullOrEmpty(entry.getSubResults())) {
            return;
        }
        for (Result subResult : entry.getSubResults()) {
            describeInternal(subResult, indent + 1);
        }
    }

    private void describeOnDepth(String message, int depth) {
        String prefix = " ".repeat((depth * 4));
        log.info(prefix + message);
    }
}

File: src/main/kotlin/com/morak/performancetracker/description/LoggingDescriptor.kt
package com.morak.performancetracker.description

import com.morak.performancetracker.context.Result
import com.morak.performancetracker.ContextType
import mu.NamedKLogging
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Component


@Component
@ConditionalOnProperty(value = ["com.morak.performance-tracker.format"], havingValue = "log", matchIfMissing = true)
class LoggingDescriptor : Descriptor {
    companion object : NamedKLogging("PERFORMANCE")

    override fun describe(root: Result, contextType: ContextType) {
        logger.info("======================{}=======================", contextType.name)
        describeInternal(root, 0)
    }

    private fun describeInternal(entry: Result, indent: Int) {
        describeOnDepth(entry.result, indent)
        entry.subResults.forEach { describeInternal(it, indent + 1) }
    }

    private fun describeOnDepth(message: String, depth: Int) {
        val prefix = " ".repeat(depth * 4)
        logger.info(prefix + message)
    }
}
